pkg/apis/seedmanagement/helper/helper.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helper
+
+import (
+	"fmt"
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/versioning"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	utilruntime.Must(config.AddToScheme(scheme))
+	utilruntime.Must(configv1alpha1.AddToScheme(scheme))
+	utilruntime.Must(gardencore.AddToScheme(scheme))
+	utilruntime.Must(gardencorev1beta1.AddToScheme(scheme))
+}
+
+// DecodeGardenletConfig decodes the given raw extension into an internal GardenletConfiguration version.
+func DecodeGardenletConfig(rawConfig *runtime.RawExtension, withDefaults bool) (*config.GardenletConfiguration, error) {
```

Comment:
I think, the way `RawExtension` is supposed to work is to set decoded objects on the `Object` field to avoid double decoding efforts.
Can you check if `rawConfig.Object` is already set to a value of type `GardenletConfiguration` and only do the decoding work if that's not the case?

Reply:
This decoding function is really special since it has the `withDefaults` parameter. This is important in cases where we don't want defaults to be applied, e.g. when applying defaults on our type. I believe any default decoding would not do this properly (it would always apply defaults) and therefore decoding explicitly is needed, independently of whether `rawConfig.Object` is already set, for `withDefaults == false`.

Reply:
ok

Reply:
In fact, testing showed that checking for `withDefaults` is not needed, because of the way defaults are currently implemented. So I'll simply add the check as suggested and adapt the tests accordingly.